### PR TITLE
Adding eigenvalue decomposition and some ergonomic functions

### PR DESCRIPTION
**Pull request by poeface**
I composed an eigenvalue decomposition and have added it to this fork. I also added a bunch of helpful functions for general scientific programming, several operators for doing element-wise matrix operations. I also added a matrix multiplication operator.

A test for each function added has been included in linalg_tests.fut.

### CODE REVIEW

File: lib/github.com/diku-dk/linalg/linalg.fut [hunk 1]
```diff
@@ -7,6 +7,28 @@
 local module type linalg = {
   -- | The scalar type.
   type t
+
+  ------ CREATING MATRICES/VECTORS ------
```
Comment (athas) [inline lib/github.com/diku-dk/linalg/linalg.fut:11]: You need a blank like after this or the documentation comment for `veczeros` will not be picked up.

File: lib/github.com/diku-dk/linalg/linalg.fut [hunk 2]
```diff
@@ -7,6 +7,28 @@
...
+  -- | Make a ones matrix.
+  val matones : (n : i64) -> (m : i64) -> [n][m]t
+  -- | Cast scalar values up into matrices (simulated broadcasting).
+  val mat [m][n] : t -> [m][n]t
+  -- | Cast integers up into matrices (simulated broadcasting).
+  val i64 [m][n] : i64 -> [m][n]t
+  -- | Cast floating point numbers up into matrices (simulated broadcasting).
+  val f64 [m][n] : f64 -> [m][n]t
+
+  ------ CORE FUNCTIONS ------
```
Comment (athas) [inline lib/github.com/diku-dk/linalg/linalg.fut:31]: Add blank line.

File: lib/github.com/diku-dk/linalg/linalg.fut [hunk 3]
```diff
@@ -19,15 +41,75 @@ local module type linalg = {
...
+  val househess [n] : [n][n]t -> ([n][n]t, [n][n]t)
+  -- | Compute the eigenvalue decomposition of matrix.
+  -- Returns (D, V), where D is a diagonal matrix of the eigenvalues, and the columns of V are their eigenvectors.
+  val eig [n] : [n][n]t -> ([n][n]t, [n][n]t)
+  -- | Compute the square root of the input matrix.
+  val matsqrt [n] : [n][n]t -> [n][n]t
+
+  ------ MATRIX OPERATORS ------
+  -- these comes last in this file so they don't interfere with the rest of the functions
+  -- short-hand matrix-level operators
```
Comment (athas) [inline lib/github.com/diku-dk/linalg/linalg.fut:90]: Add blank line.

File: lib/github.com/diku-dk/linalg/linalg.fut [hunk 4]
```diff
@@ -45,11 +127,15 @@ module type field = {
   val **: t -> t -> t
 
   val neg: t -> t
-  val <: t -> t -> bool
+  -- val inf: t
```
Comment (athas) [inline lib/github.com/diku-dk/linalg/linalg.fut:130]: Is this intentional?

**Review by athas — CHANGES_REQUESTED** [review]
I am uncertain about the arithmetic operators (`+` etc) on matrices.  I get that they can be convenient, but it seems ugly that they work specifically on matrices, and not e.g. vectors (which are also "linear algebra objects").

Further, we hope to make them redundant by improving the language itself: https://github.com/diku-dk/futhark/pull/1747

It's not difficult for users to define such operators themselves if they want them, so I don't think they belong in the `linalg` module type itself.  I would be OK with a separate module that provides them.

File: lib/github.com/diku-dk/linalg/linalg.fut (hunk 4 above)
Comment (poeface) [reply lib/github.com/diku-dk/linalg/linalg.fut:130]: I will remove that, it was left-over from some point earlier in development but it must not have been needed.

**Comment by poeface** [comment]
I see what you mean about the arithmetic operators and agree. Once the language has that automap feature, it will also render a several other included functions irrelevant as well (matop, matunary, matcomp, etc.).

Any idea what the time table is for the automap feature being added to the compiler? If it's going to happen soon then probably it wouldn't be worth adding a new module.

**Comment by athas** [comment]
Probably not *that* soon.  Hopefully this year.  There are some outstanding research questions to answer (unless we decide to go with a partial solution in the interim).

**Comment by poeface** [comment]
My initial idea for how to add this temporary module would be to make a new file which wraps linalg, adding these operators in addition to the core functions within linalg. But since linalg's module type is local, I don't think I can reexport that. Thoughts?

One other point is that since linalg already is local and the annotation there is
```
Note that this module type is declared `local`, which means you cannot directly reference
it by name from outside.  This is because it is not a stable
interface, as we may add new members in minor versions.
```
Maybe it's OK to temporarily add these utility functions in this version and then remove them later when they become irrelevant? Presumably linalg's module type will become finalized some time in the future, but for now it's already going to be changing.

I don't know the right answer, but if we know these operators will become deprecated and removed in the future, then in a way it may not matter where we put them.

**Comment by poeface** [comment]
After discussing with my colleague, I am of the opinion that we just exclude those operators and make due until automap is ready. If that is alright with you then we should be set!

**Comment by athas** [comment]
That is what I would prefer.

File: lib/github.com/diku-dk/linalg/linalg.fut [hunk 5]
```diff
@@ -89,6 +196,46 @@ module mk_linalg (T: field): linalg with t = T.t = {
   def matvecmul_col [n][m] (xss: [n][m]t) (ys: [n]t) =
     matmul xss (replicate m ys)
 
+  --UNLESS WE NOINLINE, THIS CAUSES COMPILATION TO HANG!!!
```
[the remark this answers is not in the document]
Comment (athas) [reply lib/github.com/diku-dk/linalg/linalg.fut:175]: I think it is fixed in the compiler now.  Could you check?  If so, remove the `noinline`.

File: lib/github.com/diku-dk/linalg/linalg_tests.fut [hunk 6]
```diff
@@ -2,77 +2,365 @@
 
 import "linalg"
 
-module linalg_i32 = mk_linalg { open i32 def fma a b c : i32 = a * b + c }
-module linalg_f32 = mk_linalg f32
+-- module linalg_i32 = mk_linalg { open i32 def fma a b c : i32 = a * b + c }
+-- module linalg_f32 = mk_linalg f32
```
Comment (athas) [inline lib/github.com/diku-dk/linalg/linalg_tests.fut:6]: Remove these lines.

File: lib/github.com/diku-dk/linalg/linalg.fut (hunk 1 above)
Comment (poeface) [reply lib/github.com/diku-dk/linalg/linalg.fut:11]: Is this fixed now or did I misunderstand?

File: lib/github.com/diku-dk/linalg/linalg_tests.fut (hunk 6 above)
Comment (poeface) [reply lib/github.com/diku-dk/linalg/linalg_tests.fut]: ok so removed.

File: lib/github.com/diku-dk/linalg/linalg.fut (hunk 5 above)
[the remark this answers is not in the document]
Comment (poeface) [reply lib/github.com/diku-dk/linalg/linalg.fut]: Ok agreed that appears to be fixed in my use case so I've removed the noinline flag.

**Comment by poeface** [comment]
It's true that I do have a lot of experience doing scientific programming in matlab and numpy, so I know what functions are useful in that context... However, doing all of that in the futhark context is different enough that I don't want to claim these are definitely the best choice, only that they imitate what's out there in imperative-land. I think they are a good start, and they have certainly  been useful to me already.

**Pull request merged**